internal/services/deviceregistry/device_registry_asset_resource.go

```diff
@@ -0,0 +1,743 @@
+			}
+
+			if config.SerialNumber != "" {
+				param.Properties.SerialNumber = pointer.To(config.SerialNumber)
+			}
+
+			if config.Attributes != nil {
+				param.Properties.Attributes = pointer.To(config.Attributes)
+			}
+
+			if config.DiscoveredAssetRefs != nil {
+				param.Properties.DiscoveredAssetRefs = pointer.To(config.DiscoveredAssetRefs)
+			}
+
+			if config.DefaultDatasetsConfiguration != "" {
+				param.Properties.DefaultDatasetsConfiguration = pointer.To(config.DefaultDatasetsConfiguration)
+			}
+
+			if config.DefaultEventsConfiguration != "" {
+				param.Properties.DefaultEventsConfiguration = pointer.To(config.DefaultEventsConfiguration)
+			}
+
+			param.Properties.DefaultTopic = toAzureTopic(config.DefaultTopicPath, config.DefaultTopicRetain)
+
+			param.Properties.Datasets = toAzureDatasets(config.Datasets)
+
+			param.Properties.Events = toAzureEvents(config.Events)
+
+			if err := client.CreateOrReplaceThenPoll(ctx, id, param); err != nil {
+				return fmt.Errorf("creating %s: %+v", id, err)
+			}
+
+			metadata.SetID(id)
+			return nil
+		},
+	}
+}
+
+func (r AssetResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.DeviceRegistry.AssetClient
+
+			id, err := assets.ParseAssetID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			var config AssetResourceModel
+			if err := metadata.Decode(&config); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			// Change the properties that can be updated
+			param := assets.AssetUpdate{
+				Properties: &assets.AssetUpdateProperties{},
+			}
+
+			if metadata.ResourceData.HasChange("tags") {
+				param.Tags = pointer.To(config.Tags)
+			}
+
+			if metadata.ResourceData.HasChange("attributes") {
+				param.Properties.Attributes = pointer.To(config.Attributes)
+			}
+
+			if metadata.ResourceData.HasChange("datasets") {
+				param.Properties.Datasets = toAzureDatasets(config.Datasets)
+			}
+
+			if metadata.ResourceData.HasChange("default_datasets_configuration") {
+				param.Properties.DefaultDatasetsConfiguration = pointer.To(config.DefaultDatasetsConfiguration)
+			}
+
+			if metadata.ResourceData.HasChange("default_events_configuration") {
+				param.Properties.DefaultEventsConfiguration = pointer.To(config.DefaultEventsConfiguration)
+			}
+
+			defaultTopicPathChanged := metadata.ResourceData.HasChange("default_topic_path")
+			defaultTopicRetainChanged := metadata.ResourceData.HasChange("default_topic_retain")
+			if defaultTopicPathChanged || defaultTopicRetainChanged {
+				param.Properties.DefaultTopic = &assets.TopicUpdate{}
+				if defaultTopicPathChanged {
+					param.Properties.DefaultTopic.Path = pointer.To(config.DefaultTopicPath)
+				}
+				if defaultTopicRetainChanged {
+					// Bug with `go-azure-sdk` library: you can't set retain to null because empty string will cause
```

Comment:
Linking issue https://github.com/hashicorp/go-azure-sdk/issues/1143 and PR https://github.com/hashicorp/pandora/pull/4633

The go sdk does not support sending null properties. After the PR goes in, this issue can be fixed but might require a code change here